File: packages/core/feature-flags/chain-provider.ts
import type { Decision, EvalContext, Provider } from './types';

export class ChainProvider implements Provider {
  readonly name = 'chain';
  private readonly providers: ReadonlyArray<Provider>;

  constructor(providers: ReadonlyArray<Provider | null | undefined>) {
    this.providers = providers.filter((p): p is Provider => p != null);
  }

  lookup(key: string, ctx: EvalContext): Decision | undefined {
    for (const p of this.providers) {
      const d = p.lookup(key, ctx);
      if (d !== undefined) return d;
    }
    return undefined;
  }
}

File: server/pkg/featureflag/chain_provider.go
package featureflag

import "context"

type ChainProvider struct {
	providers []Provider
}

func NewChainProvider(providers ...Provider) *ChainProvider {
	cp := &ChainProvider{providers: make([]Provider, 0, len(providers))}
	for _, p := range providers {
		if p != nil {
			cp.providers = append(cp.providers, p)
		}
	}
	return cp
}

func (*ChainProvider) Name() string { return "chain" }

func (cp *ChainProvider) Lookup(ctx context.Context, key string) (Decision, bool) {
	for _, p := range cp.providers {
		if d, ok := p.Lookup(ctx, key); ok {
			return d, true
		}
	}
	return Decision{}, false
}

func (cp *ChainProvider) Providers() []Provider {
	out := make([]Provider, len(cp.providers))
	copy(out, cp.providers)
	return out
}
